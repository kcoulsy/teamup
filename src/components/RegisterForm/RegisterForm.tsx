import React from 'react';
import { useHistory } from 'react-router-dom';
import { passwordSchema } from '../../../server/src/validation/password';
import {
  MIN_USERNAME_LENGTH,
  MIN_PASS_LENGTH,
} from './../../constants/register';
import { Form, Card, Input, Button } from 'antd';
import { PATH_LOGIN } from '../../constants/pageRoutes';
import useUser from '../../hooks/useUser';
import { ZodError } from 'zod';

export const RegisterForm = () => {
  const { register, registerError, isRegistering } = useUser();
  const history = useHistory();

  return (
    <Card title='Register' size='small' className='register-form__card'>
      <Form
        onFinish={({ username, email, password, confirm }) => {
          register({ username, email, password });
        }}>
        <Form.Item
          label='Username'
          name='username'
          rules={[
            {
              min: MIN_USERNAME_LENGTH,
              message: `Username must be at least ${MIN_USERNAME_LENGTH} characters`,
            },
            {
              required: true,
              message: 'Please input your username!',
            },
          ]}>
          <Input />
        </Form.Item>
        <Form.Item
          label='Email'
          name='email'
          rules={[
            {
              type: 'email',
              message: 'The input is not valid email!',
            },
            {
              required: true,
              message: 'Please input an email!',
            },
          ]}>
          <Input />
        </Form.Item>
        <Form.Item
          label='Password'
          name='password'
          rules={[
            {
              required: true,
              message: 'Please input your password!',
            },
            () => ({
              validator(_, value) {
                if (value === '') {
                  return Promise.resolve();
                }
                try {
                  passwordSchema.parse(value);
                  return Promise.resolve();
                } catch (err) {
                  if (err instanceof ZodError) {
                    return Promise.reject(
                      err?.errors.map(({ message }) => message).join('\r\n') +
                        '.'
                    );
                  }
                  return Promise.reject('Invalid password.');
                }
              },
            }),
          ]}>
          <Input.Password />
        </Form.Item>
        <Form.Item
          label='Confirm'
          name='confirm'
          rules={[
            {
              required: true,
              message: 'Please confirm your password!',
            },
            ({ getFieldValue }) => ({
              validator(_, value) {
                if (!value || getFieldValue('password') === value) {
                  return Promise.resolve();
                }
                return Promise.reject(
                  'The two passwords that you entered do not match!'
                );
              },
            }),
          ]}>
          <Input.Password />
        </Form.Item>
        {registerError ? (
          <Form.Item>Something went wrong trying to register you</Form.Item>
        ) : null}
        <Form.Item>
          <Button type='primary' htmlType='submit' loading={isRegistering}>
            Register
          </Button>
          <Button
            type='link'
            htmlType='button'
            onClick={() => history.push(PATH_LOGIN)}>
            Already registered? Click here to login
          </Button>
        </Form.Item>
      </Form>
    </Card>
  );
};

export default RegisterForm;
