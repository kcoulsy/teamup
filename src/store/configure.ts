import { configureStore } from '@reduxjs/toolkit';

// Import Reducers
import authReducer from '../reducers/auth';
import teamReducer from '../reducers/team';

const store = configureStore({
    reducer: {
        // reducers here
        auth: authReducer,
        team: teamReducer,
    },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
