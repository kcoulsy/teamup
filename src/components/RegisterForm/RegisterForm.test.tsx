import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { describe, expect, test, vi } from 'vitest';
import { RegisterForm, mapStateToProps } from './RegisterForm';
import { RootState } from '../../store/configure';

describe('render login form component', () => {
  test('should render RegisterForm component', () => {
    const startRegister = vi.fn();
    const wrapper = shallow(
      <RegisterForm attemptingRegister={false} startRegister={startRegister} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  test('should call startRegister on submit', () => {
    const startRegister = vi.fn();
    const wrapper = shallow(
      <RegisterForm attemptingRegister={false} startRegister={startRegister} />
    );
    const testUser = {
      username: 'username',
      email: 'test@test.com',
      password: 'password1',
      confirm: 'password1',
    };
    wrapper
      .find('input[name="username"]')
      .simulate('change', { target: { value: testUser.username } });
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: testUser.email } });
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('input[name="confirm"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(1);
    expect(startRegister.mock.calls[0][0]).toEqual(testUser);
  });

  test('should show errors if there are password problems', () => {
    const startRegister = vi.fn();
    const wrapper = shallow(
      <RegisterForm attemptingRegister={false} startRegister={startRegister} />
    );
    const testUser = {
      username: 'username',
      email: 'test@test.com',
      password: 'password1',
      confirm: 'password1',
    };

    // MISSING PASSWORD FIELDS
    wrapper
      .find('input[name="username"]')
      .simulate('change', { target: { value: testUser.username } });
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: testUser.email } });
    // Not entering a password yet.
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(0);
    expect(wrapper.find('.error.message').length).toBe(1);

    expect(wrapper.find('li[data-error="password-empty"]').length).toBe(1);
    expect(wrapper.find('li[data-error="confirm-empty"]').length).toBe(1);

    // MISMATCHED PASSWORDS
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('input[name="confirm"]')
      .simulate('change', { target: { value: 'notthesame' } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(wrapper.find('li[data-error="password-mismatch"]').length).toBe(1);

    // INVALID PASSWORD (MIN LEN 3, CONTAINS NUMBER)
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { value: 'abc' } });
    wrapper
      .find('input[name="confirm"]')
      .simulate('change', { target: { value: 'abc' } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(wrapper.find('li[data-error="password-invalid"]').length).toBe(1);

    // VALID PASSWORD, REMOVE THE ERRORS
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('input[name="confirm"]')
      .simulate('change', { target: { value: testUser.confirm } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(1);
    expect(wrapper.find('.error.message').length).toBe(0);
  });

  test('should show errors if there are username problems', () => {
    const startRegister = vi.fn();
    const wrapper = shallow(
      <RegisterForm attemptingRegister={false} startRegister={startRegister} />
    );
    const testUser = {
      username: 'username',
      email: 'test@test.com',
      password: 'password1',
      confirm: 'password1',
    };

    // NO USERNAME ENTERED
    wrapper
      .find('input[name="username"]')
      .simulate('change', { target: { value: '' } });
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: testUser.email } });
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('input[name="confirm"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(0);
    expect(wrapper.find('.error.message').length).toBe(1);
    expect(wrapper.find('form').hasClass('error')).toBe(true);

    expect(wrapper.find('li[data-error="username-empty"]').length).toBe(1);

    // INVALID USERNAME LENGTH
    wrapper
      .find('input[name="username"]')
      .simulate('change', { target: { value: 'ab' } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(wrapper.find('.error.message').length).toBe(1);
    expect(wrapper.find('form').hasClass('error')).toBe(true);
    expect(wrapper.find('li[data-error="username-short"]').length).toBe(1);

    expect(startRegister.mock.calls.length).toBe(0);

    // VALID USERNAME
    wrapper
      .find('input[name="username"]')
      .simulate('change', { target: { value: testUser.username } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(1);
    expect(wrapper.find('.error.message').length).toBe(0);
  });

  test('should show errors if there are email problems', () => {
    const startRegister = vi.fn();
    const wrapper = shallow(
      <RegisterForm attemptingRegister={false} startRegister={startRegister} />
    );
    const testUser = {
      username: 'username',
      email: 'test@test.com',
      password: 'password1',
      confirm: 'password1',
    };

    // NO USERNAME ENTERED
    wrapper
      .find('input[name="username"]')
      .simulate('change', { target: { value: testUser.username } });
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: '' } });
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('input[name="confirm"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(0);
    expect(wrapper.find('.error.message').length).toBe(1);
    expect(wrapper.find('form').hasClass('error')).toBe(true);

    expect(wrapper.find('li[data-error="email-empty"]').length).toBe(1);

    // INVALID USERNAME LENGTH
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: 'notavalidemail' } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(wrapper.find('.error.message').length).toBe(1);
    expect(wrapper.find('form').hasClass('error')).toBe(true);
    expect(wrapper.find('li[data-error="email-invalid"]').length).toBe(1);

    expect(startRegister.mock.calls.length).toBe(0);

    // VALID USERNAME
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: testUser.email } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(1);
    expect(wrapper.find('.error.message').length).toBe(0);
  });

  test('should show errors if errorMsg is passed in as a prop', () => {
    const startRegister = vi.fn();
    const ERROR_MSG = 'Test error message';
    const wrapper = shallow(
      <RegisterForm
        attemptingRegister={false}
        startRegister={startRegister}
        errorMsg={ERROR_MSG}
      />
    );
    const testUser = {
      username: 'username',
      email: 'test@test.com',
      password: 'password1',
      confirm: 'password1',
    };

    wrapper
      .find('input[name="username"]')
      .simulate('change', { target: { value: testUser.username } });
    wrapper
      .find('input[name="email"]')
      .simulate('change', { target: { value: testUser.email } });
    wrapper
      .find('input[name="password"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('input[name="confirm"]')
      .simulate('change', { target: { value: testUser.password } });
    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(0);
    expect(wrapper.find('.error.message').length).toBe(1);
    expect(wrapper.find('form').hasClass('error')).toBe(true);

    expect(wrapper.find('li[data-error="error-msg"]').length).toBe(1);

    wrapper.setProps({ errorMsg: null });

    wrapper
      .find('button[type="submit"]')
      .simulate('click', { preventDefault: vi.fn() });

    expect(startRegister.mock.calls.length).toBe(1);
    expect(wrapper.find('.error.message').length).toBe(0);
  });

  test('should show loader with attemptingRegister', () => {
    const startRegister = vi.fn();
    const wrapper = shallow(
      <RegisterForm startRegister={startRegister} attemptingRegister={true} />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();

    expect(wrapper.find('Loader').length).toBe(1);
  });

  test('should correctly map state to props', () => {
    const mockState: RootState = {
      auth: {
        attemptingLogin: false,
        token: null,
        loginAttemptFailed: false,
        appInitialising: false,
        appInitialised: false,
        attemptingRegister: false,
        registerErrorMsg: 'error message',
      },
    };

    expect(mapStateToProps(mockState).attemptingRegister).toEqual(
      mockState.auth.attemptingRegister
    );
    expect(mapStateToProps(mockState).errorMsg).toEqual(
      mockState.auth.registerErrorMsg
    );
  });
});
