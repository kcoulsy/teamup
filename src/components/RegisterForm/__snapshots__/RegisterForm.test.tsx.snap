// Vitest Snapshot v1

exports[`render login form component > should render RegisterForm component 1`] = `
<Card
  className="register-form__card"
  size="small"
  title="Register"
>
  <ForwardRef(InternalForm)
    onFinish={[Function]}
  >
    <FormItem
      label="Username"
      name="username"
      rules={
        [
          {
            "message": "Username must be at least 3 characters",
            "min": 3,
          },
          {
            "message": "Please input your username!",
            "required": true,
          },
        ]
      }
    >
      <Input
        type="text"
      />
    </FormItem>
    <FormItem
      label="Email"
      name="email"
      rules={
        [
          {
            "message": "The input is not valid email!",
            "type": "email",
          },
          {
            "message": "Please input an email!",
            "required": true,
          },
        ]
      }
    >
      <Input
        type="text"
      />
    </FormItem>
    <FormItem
      label="Password"
      name="password"
      rules={
        [
          {
            "message": "Please input your password!",
            "required": true,
          },
          {
            "message": "Password must be at least 6 characters",
            "min": 6,
          },
          [Function],
        ]
      }
    >
      <Password
        action="click"
        iconRender={[Function]}
        visibilityToggle={true}
      />
    </FormItem>
    <FormItem
      label="Confirm"
      name="confirm"
      rules={
        [
          {
            "message": "Please confirm your password!",
            "required": true,
          },
          [Function],
        ]
      }
    >
      <Password
        action="click"
        iconRender={[Function]}
        visibilityToggle={true}
      />
    </FormItem>
    <FormItem>
      <Button
        block={false}
        ghost={false}
        htmlType="submit"
        loading={false}
        type="primary"
      >
        Register
      </Button>
      <Button
        block={false}
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
        type="link"
      >
        Already registered? Click here to login
      </Button>
    </FormItem>
  </ForwardRef(InternalForm)>
</Card>
`;

exports[`render login form component > should show loader with attemptingRegister 1`] = `
<Card
  className="register-form__card"
  size="small"
  title="Register"
>
  <ForwardRef(InternalForm)
    onFinish={[Function]}
  >
    <FormItem
      label="Username"
      name="username"
      rules={
        [
          {
            "message": "Username must be at least 3 characters",
            "min": 3,
          },
          {
            "message": "Please input your username!",
            "required": true,
          },
        ]
      }
    >
      <Input
        type="text"
      />
    </FormItem>
    <FormItem
      label="Email"
      name="email"
      rules={
        [
          {
            "message": "The input is not valid email!",
            "type": "email",
          },
          {
            "message": "Please input an email!",
            "required": true,
          },
        ]
      }
    >
      <Input
        type="text"
      />
    </FormItem>
    <FormItem
      label="Password"
      name="password"
      rules={
        [
          {
            "message": "Please input your password!",
            "required": true,
          },
          {
            "message": "Password must be at least 6 characters",
            "min": 6,
          },
          [Function],
        ]
      }
    >
      <Password
        action="click"
        iconRender={[Function]}
        visibilityToggle={true}
      />
    </FormItem>
    <FormItem
      label="Confirm"
      name="confirm"
      rules={
        [
          {
            "message": "Please confirm your password!",
            "required": true,
          },
          [Function],
        ]
      }
    >
      <Password
        action="click"
        iconRender={[Function]}
        visibilityToggle={true}
      />
    </FormItem>
    <FormItem>
      <Button
        block={false}
        ghost={false}
        htmlType="submit"
        loading={true}
        type="primary"
      >
        Register
      </Button>
      <Button
        block={false}
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
        type="link"
      >
        Already registered? Click here to login
      </Button>
    </FormItem>
  </ForwardRef(InternalForm)>
</Card>
`;

exports[`render login form component should render RegisterForm component 1`] = `
<Card
  centered={true}
  title="Register"
>
  <Loader
    isLoading={false}
    loadingText="Registering"
  >
    <form
      className="ui form"
    >
      <div
        className="field"
      >
        <label>
          Username
        </label>
        <input
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Email
        </label>
        <input
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Password
        </label>
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Confirm Password
        </label>
        <input
          name="confirm"
          onChange={[Function]}
          placeholder="Confirm Password"
          type="password"
        />
      </div>
      <div
        className="field"
      >
        <div
          className="ui label"
        >
          Passwords must be at least 
          6
           
          characters long and contain a number.
        </div>
      </div>
      <button
        className="ui button"
        onClick={[Function]}
        type="submit"
      >
        Register
      </button>
      Already registered? 
      <Link
        to="/login"
      >
        Login
      </Link>
    </form>
  </Loader>
</Card>
`;

exports[`render login form component should show loader with attemptingRegister 1`] = `
<Card
  centered={true}
  title="Register"
>
  <Loader
    isLoading={true}
    loadingText="Registering"
  >
    <form
      className="ui form"
    >
      <div
        className="field"
      >
        <label>
          Username
        </label>
        <input
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Email
        </label>
        <input
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Password
        </label>
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Confirm Password
        </label>
        <input
          name="confirm"
          onChange={[Function]}
          placeholder="Confirm Password"
          type="password"
        />
      </div>
      <div
        className="field"
      >
        <div
          className="ui label"
        >
          Passwords must be at least 
          6
           
          characters long and contain a number.
        </div>
      </div>
      <button
        className="ui button"
        onClick={[Function]}
        type="submit"
      >
        Register
      </button>
      Already registered? 
      <Link
        to="/login"
      >
        Login
      </Link>
    </form>
  </Loader>
</Card>
`;
